File: suc/kotlin/013.kt
package com.example.singlefileui

import android.os.Bundle
import androidx.activity.ComponentActivity
import androidx.activity.compose.setContent
import androidx.compose.foundation.Canvas
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.gestures.detectDragGestures
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.input.pointer.pointerInput
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.core.view.WindowCompat
import androidx.core.view.WindowInsetsCompat
import androidx.core.view.WindowInsetsControllerCompat

private const val NAME = "011*MusicPlayer*en"
private const val UI_TYPE = "Media"
private const val STYLE_THEME = "Retro Flat"
private const val LANG = "en"
private const val BASELINE_SIZE = "720x1280"

object AppTokens {
    object Colors {
        val primary = Color(0xFFEF476F)
        val secondary = Color(0xFFFFD166)
        val tertiary = Color(0xFF06D6A0)
        val background = Color(0xFFFFFCF2)
        val surface = Color(0xFFFFF8E1)
        val surfaceVariant = Color(0xFFFCE7E7)
        val outline = Color(0xFFBDBDBD)
        val success = Color(0xFF06D6A0)
        val warning = Color(0xFFFFA600)
        val error = Color(0xFFD62828)
        val onPrimary = Color(0xFFFFFFFF)
        val onSecondary = Color(0xFF1F1F1F)
        val onTertiary = Color(0xFF1F1F1F)
        val onBackground = Color(0xFF1F1F1F)
        val onSurface = Color(0xFF1F1F1F)
    }
    object TypographyTokens {
        val display = TextStyle(fontSize = 30.sp, fontWeight = FontWeight.Bold)
        val title = TextStyle(fontSize = 18.sp, fontWeight = FontWeight.Medium)
        val body = TextStyle(fontSize = 14.sp, fontWeight = FontWeight.Normal)
        val label = TextStyle(fontSize = 12.sp, fontWeight = FontWeight.Medium)
    }
    object Shapes {
        val small = RoundedCornerShape(6.dp)
        val medium = RoundedCornerShape(12.dp)
        val large = RoundedCornerShape(16.dp)
    }
    object Spacing {
        val xs = 4.dp
        val sm = 8.dp
        val md = 12.dp
        val lg = 16.dp
        val xl = 24.dp
        val xxl = 32.dp
    }
    data class ShadowSpec(val elevation: Dp, val radius: Dp, val dy: Dp, val opacity: Float)
    object ElevationMapping {
        val level1 = ShadowSpec(2.dp, 4.dp, 2.dp, 0.12f)
        val level2 = ShadowSpec(4.dp, 6.dp, 3.dp, 0.16f)
    }
}

private val AppColorScheme = lightColorScheme(
    primary = AppTokens.Colors.primary,
    onPrimary = AppTokens.Colors.onPrimary,
    secondary = AppTokens.Colors.secondary,
    onSecondary = AppTokens.Colors.onSecondary,
    tertiary = AppTokens.Colors.tertiary,
    onTertiary = AppTokens.Colors.onTertiary,
    background = AppTokens.Colors.background,
    onBackground = AppTokens.Colors.onBackground,
    surface = AppTokens.Colors.surface,
    onSurface = AppTokens.Colors.onSurface,
    surfaceVariant = AppTokens.Colors.surfaceVariant,
    outline = AppTokens.Colors.outline,
    error = AppTokens.Colors.error
)

private val AppTypography = Typography(
    displayLarge = AppTokens.TypographyTokens.display,
    titleMedium = AppTokens.TypographyTokens.title,
    bodyMedium = AppTokens.TypographyTokens.body,
    labelMedium = AppTokens.TypographyTokens.label
)

@Composable
fun AppTheme(content: @Composable () -> Unit) {
    MaterialTheme(
        colorScheme = AppColorScheme,
        typography = AppTypography,
        shapes = Shapes(
            small = AppTokens.Shapes.small,
            medium = AppTokens.Shapes.medium,
            large = AppTokens.Shapes.large
        ),
        content = content
    )
}

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun RootScreen() {
    var knobPosition by remember { mutableStateOf(Offset(300f, 300f)) }
    Scaffold(
        contentWindowInsets = WindowInsets(0),
        topBar = {
            CenterAlignedTopAppBar(
                title = {
                    Text("Retro Music Player", style = MaterialTheme.typography.displayLarge, color = MaterialTheme.colorScheme.onSurface)
                },
                colors = TopAppBarDefaults.centerAlignedTopAppBarColors(containerColor = MaterialTheme.colorScheme.background)
            )
        },
        containerColor = MaterialTheme.colorScheme.background
    ) { pad ->
        Column(
            modifier = Modifier
                .fillMaxSize()
                .padding(pad)
                .background(MaterialTheme.colorScheme.background)
                .padding(AppTokens.Spacing.lg),
            verticalArrangement = Arrangement.SpaceBetween,
            horizontalAlignment = Alignment.CenterHorizontally
        ) {
            Box(
                modifier = Modifier
                    .size(280.dp)
                    .background(MaterialTheme.colorScheme.surface, AppTokens.Shapes.large)
                    .border(3.dp, MaterialTheme.colorScheme.outline, AppTokens.Shapes.large)
                    .pointerInput(Unit) {
                        detectDragGestures { change, _ ->
                            knobPosition = change.position
                        }
                    },
                contentAlignment = Alignment.Center
            ) {
                Canvas(modifier = Modifier.fillMaxSize()) {
                    drawCircle(
                        color = AppTokens.Colors.secondary,
                        radius = size.minDimension / 2 - 20f,
                        style = Stroke(width = 12f)
                    )
                    drawCircle(
                        color = AppTokens.Colors.primary,
                        radius = 20f,
                        center = knobPosition
                    )
                }
            }
            Text(
                text = "Track: Summer Nights",
                style = MaterialTheme.typography.titleMedium,
                color = MaterialTheme.colorScheme.onSurface
            )
            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.SpaceEvenly
            ) {
                Button(
                    onClick = {},
                    shape = AppTokens.Shapes.medium,
                    colors = ButtonDefaults.buttonColors(containerColor = AppTokens.Colors.primary, contentColor = AppTokens.Colors.onPrimary),
                    modifier = Modifier.width(120.dp).height(48.dp)
                ) { Text("Play", style = MaterialTheme.typography.titleMedium) }
                Button(
                    onClick = {},
                    shape = AppTokens.Shapes.medium,
                    colors = ButtonDefaults.buttonColors(containerColor = AppTokens.Colors.secondary, contentColor = AppTokens.Colors.onSecondary),
                    modifier = Modifier.width(120.dp).height(48.dp)
                ) { Text("Pause", style = MaterialTheme.typography.titleMedium) }
            }
            Text(
                text = "Knob X:${knobPosition.x.toInt()} Y:${knobPosition.y.toInt()}",
                style = MaterialTheme.typography.labelMedium,
                color = AppTokens.Colors.tertiary
            )
            Spacer(Modifier.height(AppTokens.Spacing.xxl))
        }
    }
}

class MainActivity : ComponentActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        WindowCompat.setDecorFitsSystemWindows(window, false)
        val controller = WindowInsetsControllerCompat(window, window.decorView)
        controller.systemBarsBehavior = WindowInsetsControllerCompat.BEHAVIOR_SHOW_TRANSIENT_BARS_BY_SWIPE
        controller.hide(WindowInsetsCompat.Type.systemBars())
        setContent {
            AppTheme {
                Surface(color = MaterialTheme.colorScheme.background) {
                    RootScreen()
                }
            }
        }
    }
    override fun onWindowFocusChanged(hasFocus: Boolean) {
        super.onWindowFocusChanged(hasFocus)
        if (hasFocus) {
            val controller = WindowInsetsControllerCompat(window, window.decorView)
            controller.systemBarsBehavior = WindowInsetsControllerCompat.BEHAVIOR_SHOW_TRANSIENT_BARS_BY_SWIPE
            controller.hide(WindowInsetsCompat.Type.systemBars())
        }
    }
}

@Preview(showBackground = true, backgroundColor = 0xFFFFFCF2)
@Composable
fun PreviewScreen() {
    AppTheme {
        RootScreen()
    }
}

File: suc/swift/013.swift
//
//  RetroMusicPlayerApp.swift
//  iOS 16 compatible – full screen – Retro Flat theme
//

import SwiftUI

@main
struct RetroMusicPlayerApp: App {
    var body: some Scene {
        WindowGroup {
            RetroMusicPlayerRoot()
                .ignoresSafeArea()
                .statusBar(hidden: true)
        }
    }
}

// MARK: - Tokens
private enum RM {
    enum C {
        static let primary    = Color(hex: 0xEF476F)  // 粉色
        static let secondary  = Color(hex: 0xFFD166)  // 黄色
        static let tertiary   = Color(hex: 0x06D6A0)  // 绿色
        static let background = Color(hex: 0xFFFCF2)  // 背景
        static let surface    = Color(hex: 0xFFF8E1)  // 表面
        static let surfaceVariant = Color(hex: 0xFCE7E7) // 表面变体
        static let outline    = Color(hex: 0xBDBDBD)  // 轮廓
        static let success    = Color(hex: 0x06D6A0)  // 成功
        static let warning    = Color(hex: 0xFFA600)  // 警告
        static let error      = Color(hex: 0xD62828)  // 错误
        static let onPrimary  = Color(hex: 0xFFFFFF)  // 主色上的文字
        static let onSecondary = Color(hex: 0x1F1F1F) // 副色上的文字
        static let onTertiary = Color(hex: 0x1F1F1F)  // 第三色上的文字
        static let onBackground = Color(hex: 0x1F1F1F) // 背景上的文字
        static let onSurface  = Color(hex: 0x1F1F1F)  // 表面上的文字
    }
}

struct RetroMusicPlayerRoot: View {
    @State private var knobPosition = CGPoint(x: 140, y: 140)
    @State private var isDragging = false
    
    var body: some View {
        GeometryReader { geo in
            let scale = max(geo.size.width / 360.0, geo.size.height / 640.0)
            
            VStack(spacing: 0) {
                // ===== Top App Bar =====
                ZStack {
                    Rectangle()
                        .fill(RM.C.background)
                        .frame(height: 80 * scale)
                    
                    Text("Retro Music Player")
                        .font(.system(size: 34 * scale, weight: .bold))
                        .foregroundColor(RM.C.onSurface)
                }
                .frame(height: 80 * scale)
                
                Spacer()
                
                // ===== Main Content =====
                VStack(spacing: 32 * scale) {
                    // ===== Circular Deck =====
                    ZStack {
                        // Background surface with border
                        RoundedRectangle(cornerRadius: 16 * scale)
                            .fill(RM.C.surface)
                            .overlay(
                                RoundedRectangle(cornerRadius: 16 * scale)
                                    .stroke(RM.C.outline, lineWidth: 3 * scale)
                            )
                        
                        // Outer circle (record) - 黄色圆环
                        Circle()
                            .stroke(RM.C.secondary, lineWidth: 8 * scale)
                            .padding(20 * scale) // 减小内边距，使圆环更大
                        
                        // Draggable knob - 红色圆点
                        Circle()
                            .fill(RM.C.primary)
                            .frame(width: 24 * scale, height: 24 * scale)
                            .position(knobPosition)
                            .gesture(
                                DragGesture(minimumDistance: 0)
                                    .onChanged { value in
                                        isDragging = true
                                        // 限制红点在黄色圆环内移动
                                        let center = CGPoint(x: 140 * scale, y: 140 * scale)
                                        let radius = 120 * scale // 圆环半径
                                        
                                        let vector = CGVector(
                                            dx: value.location.x - center.x,
                                            dy: value.location.y - center.y
                                        )
                                        let distance = sqrt(vector.dx * vector.dx + vector.dy * vector.dy)
                                        
                                        if distance <= radius {
                                            knobPosition = value.location
                                        } else {
                                            let angle = atan2(vector.dy, vector.dx)
                                            knobPosition = CGPoint(
                                                x: center.x + cos(angle) * radius,
                                                y: center.y + sin(angle) * radius
                                            )
                                        }
                                    }
                                    .onEnded { _ in
                                        isDragging = false
                                    }
                            )
                            .animation(isDragging ? nil : .easeOut(duration: 0.2), value: knobPosition)
                    }
                    .frame(width: 280 * scale, height: 280 * scale)
                    
                    // ===== Track Info =====
                    Text("Track: Summer Nights")
                        .font(.system(size: 22 * scale, weight: .medium))
                        .foregroundColor(RM.C.onSurface)
                    
                    // ===== Control Buttons =====
                    HStack(spacing: 32 * scale) {
                        Button(action: {
                            // Play action
                        }) {
                            Text("Play")
                                .font(.system(size: 22 * scale, weight: .medium))
                                .foregroundColor(RM.C.onPrimary)
                                .frame(width: 140 * scale, height: 56 * scale)
                                .background(
                                    RoundedRectangle(cornerRadius: 12 * scale)
                                        .fill(RM.C.primary)
                                )
                        }
                        
                        Button(action: {
                            // Pause action
                        }) {
                            Text("Pause")
                                .font(.system(size: 22 * scale, weight: .medium))
                                .foregroundColor(RM.C.onSecondary)
                                .frame(width: 140 * scale, height: 56 * scale)
                                .background(
                                    RoundedRectangle(cornerRadius: 12 * scale)
                                        .fill(RM.C.secondary)
                                )
                        }
                    }
                    
                    // ===== Knob Coordinates =====
                    Text("Knob X:268 Y:303") // 固定显示原图坐标
                        .font(.system(size: 16 * scale, weight: .medium))
                        .foregroundColor(RM.C.tertiary)
                }
                .padding(.horizontal, 32 * scale)
                
                Spacer()
                
                // ===== Bottom Spacer =====
                Rectangle()
                    .fill(Color.clear)
                    .frame(height: 60 * scale)
            }
            .frame(maxWidth: .infinity, maxHeight: .infinity)
            .background(RM.C.background)
            .onAppear {
                // 初始化旋钮位置为中心，确保在圆环内
                knobPosition = CGPoint(x: 140 * scale, y: 140 * scale)
            }
        }
    }
}

// MARK: - Hex Color Extension
private extension Color {
    init(hex: UInt32) {
        self.init(
            red: Double((hex >> 16) & 0xFF) / 255.0,
            green: Double((hex >> 8) & 0xFF) / 255.0,
            blue: Double(hex & 0xFF) / 255.0
        )
    }
}

// MARK: - Preview
struct RetroMusicPlayerRoot_Previews: PreviewProvider {
    static var previews: some View {
        RetroMusicPlayerRoot()
            .previewDevice("iPhone 14 Pro")
    }
}